src/theory/quantifiers/sygus_grammar_norm.cpp

```diff
@@ -17,162 +17,472 @@
-      tos[i].d_ops.push_back(Rewriter::rewrite(exp_sop_n));
-      Trace("sygus-grammar-normalize")
-          << "\tOriginal op: " << cons.getSygusOp()
-          << "\n\tExpanded one: " << exp_sop_n
-          << "\n\tRewritten one: " << tos[i].d_ops.back() << std::endl;
-      tos[i].d_cons_names.push_back(cons.getName());
-      tos[i].d_pcb.push_back(cons.getSygusPrintCallback());
-      tos[i].d_cons_args_t.push_back(std::vector<Type>());
-      for (const DatatypeConstructorArg& arg : cons)
+  TypeNode sygus_tn = TypeNode::fromType(dt.getSygusType());
+  /* TODO #1304: step 0: look for singleton */
+  /* TODO #1304: step 1: look for redundant constructors to drop */
+  /* step 2: look for chain */
+  unsigned chain_op_pos = dt.getNumConstructors();
+  std::vector<unsigned> elem_pos;
+  for (unsigned i = 0, size = op_pos.size(); i < size; ++i)
+  {
+    Assert(op_pos[i] < dt.getNumConstructors());
+    Expr sop = dt[op_pos[i]].getSygusOp();
+    /* Collects a chainable operator such as PLUS */
+    if (sop.getKind() == BUILTIN
+        && TransfChain::isChainable(sygus_tn, Node::fromExpr(sop)))
+    {
+      Assert(nm->operatorToKind(Node::fromExpr(sop)) == PLUS);
+      /* TODO #1304: be robust for this case */
+      /* For now only transforms applications whose arguments have the same type
+       * as the root */
+      bool same_type_plus = true;
+      for (const DatatypeConstructorArg& arg : dt[op_pos[i]])
       {
-        /* Collect unresolved types corresponding to the typenode of the
-         * arguments */
-        tos[i].d_cons_args_t.back().push_back(tn_to_unres[TypeNode::fromType(
-            static_cast<SelectorType>(arg.getType()).getRangeType())]);
+        if (TypeNode::fromType(
+                static_cast<SelectorType>(arg.getType()).getRangeType())
+            != tn)
+        {
+          same_type_plus = false;
+          break;
+        }
       }
+      if (!same_type_plus)
+      {
+        Trace("sygus-grammar-normalize-infer")
+            << "\tFor OP " << PLUS << " did not collecting sop " << sop
+            << " in position " << op_pos[i] << "\n";
+        continue;
+      }
+      Assert(chain_op_pos == dt.getNumConstructors());
+      Trace("sygus-grammar-normalize-infer")
+          << "\tCollecting chainable OP " << sop << " in position " << op_pos[i]
+          << "\n";
+      chain_op_pos = op_pos[i];
+      continue;
     }
-    /* Use the sygus type to not lose reference to the original types (Bool,
-     * Int, etc) */
-    tos[i].d_dt.setSygus(dt.getSygusType(),
-                         sygus_vars.toExpr(),
-                         dt.getSygusAllowConst(),
-                         dt.getSygusAllowAll());
-    for (unsigned j = 0, size_d_ops = tos[i].d_ops.size(); j < size_d_ops; ++j)
+    /* TODO #1304: check this for each operator */
+    /* Collects elements different from the identity (e.g. zero) */
+    Kind actual_sop_k = nm->operatorToKind(sop);
+    if (sop.getKind() == BOUND_VARIABLE
```

Comment:
This condition can be simplified to just !isId right?

Reply:
Not as it is because I'm collecting everything that is not an identity of plus (so e.g. MINUS would be collected).

Reply:
Yes, so !isId( MINUS ) holds, so this would work, right?